tests/cxx/typeid.cc

```diff
@@ -0,0 +1,37 @@
+//===--- typeid.cc - test input file for iwyu -----------------------------===//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+//===----------------------------------------------------------------------===//
+
+// IWYU_ARGS: -I .
+
+// Tests that IWYU suggests an include instead of a forward declaration for
+// expressions used as arguments to typeid(). This is necessary as calling
+// typeid() on an incomplete type causes the program to be ill-formed.
+
+#include "tests/cxx/direct.h"
+#include <typeinfo>
+
+// IWYU: IndirectClass needs a declaration
+void GetTypeInfoOfIndirectClass(const IndirectClass &indirect) {
+  // IWYU: IndirectClass is...*indirect.h
+  const std::type_info &typeInfo = typeid(indirect);
```

Comment:
I'm guessing this only covers the type-operand path. Would be nice to see one for the expr-operand as well!